proving_system/stark/src/air/frame.rs

```diff
@@ -25,30 +29,36 @@ impl<F: IsFFTField> Frame<F> {
         self.row_width
     }
 
-    pub fn get_row(&self, row_idx: usize) -> &[FieldElement<F>] {
+    pub fn get_row(&self, row_idx: usize) -> Result<&[FieldElement<F>], AIRError> {
         let row_offset = row_idx * self.row_width;
-        &self.data[row_offset..row_offset + self.row_width]
+        self.data
+            .get(row_offset..row_offset + self.row_width)
+            .ok_or(AIRError::RowIndexOutOfFrameBounds(row_idx, self.row_width))
     }
 
-    pub fn get_row_mut(&mut self, row_idx: usize) -> &mut [FieldElement<F>] {
+    pub fn get_row_mut(&mut self, row_idx: usize) -> Result<&mut [FieldElement<F>], AIRError> {
         let row_offset = row_idx * self.row_width;
-        &mut self.data[row_offset..row_offset + self.row_width]
+        self.data
+            .get_mut(row_offset..row_offset + self.row_width)
+            .ok_or(AIRError::RowIndexOutOfFrameBounds(row_idx, self.row_width))
     }
 
     pub fn read_from_trace(
         trace: &TraceTable<F>,
         step: usize,
         blowup: u8,
         offsets: &[usize],
-    ) -> Self {
+    ) -> Result<Self, AIRError> {
         // Get trace length to apply module with it when getting elements of
         // the frame from the trace.
         let trace_steps = trace.n_rows();
         let data = offsets
             .iter()
             .flat_map(|frame_row_idx| {
+                // It's ok to unwrap here because the row index will always be in bounds
```

Comment:
I think here we are assuming that `self.data`'s rows will always have the same length (i.e. its length is always divisible by `n_rows`). Is that ok?